Fully responsive globals.php with vertical menu

zbig01 opened the pull request:
Replaced tabs with vertical left menu for globals.php

Now has a workable UI on iPad and Smartphones

Some screenshots

**Default view condensed:**

![edit_globals-01](https://user-images.githubusercontent.com/12654295/58618079-86c01d80-8276-11e9-9954-8cd37c35d67a.png)



**Default view full screen:**

![edit_globals-02](https://user-images.githubusercontent.com/12654295/58618095-93dd0c80-8276-11e9-9c70-fe6c7f6b2a90.png)



**User Settings:**

![edit_globals-03](https://user-images.githubusercontent.com/12654295/58618150-bbcc7000-8276-11e9-8068-f38608c353c3.png)



**On Tablet - landscape:**

![edit_globals-04](https://user-images.githubusercontent.com/12654295/58618177-cab32280-8276-11e9-9ebc-22e9ba6d7950.png)



**On Smartphone - Portrait:**

![edit_globals-05](https://user-images.githubusercontent.com/12654295/58618204-def71f80-8276-11e9-94e8-9150604c11be.png)



**On Smartphone - Landscape:**

![edit_globals-06](https://user-images.githubusercontent.com/12654295/58618227-eb7b7800-8276-11e9-93e2-6754a6201a8b.png)




below added by bradymiller:
Up For Grabs demo for this PR is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Zeta_-_Up_For_Grabs_Demo

adunsulag commented:
@zbig01 I just checked this out to my local repo and the main column is being cleared to the bottom.  I'm running google chrome 74 on Ubuntu 16.04 and I get the same result on firefox 67.  Caches have been cleared, etc.

![image](https://user-images.githubusercontent.com/228117/58673741-baed1a00-831a-11e9-98ec-4c3fa7b0c2ef.png)

It appears that changing the styles for div.tabContainer to the following fixes the issues
```
div.tabContainer {
    clear: right; /** was clear: both **/
    float: left;
    width: 90%; /** was 100% **/
```
Since tabNav is set to 10% and both columns are floated everything was being forced to the bottom.
Perhaps I haven't set the branch up correctly or there is some other odd behavior on why it's not showing up correctly for me but this was the first thing I saw.

adunsulag commented:
@zbig01 I should mention that I'm running off the docker so perhaps there is a different environment there?

zbig01 commented:
@adunsulag  did you run `npm run build` to recompile the stylesheets and clear your cache ? The stylesheets have been reworked .  I note from your screenshot that the light color theme vertical menu has solid blue menu. this is how it should look

![edit_globals_light](https://user-images.githubusercontent.com/12654295/58674211-a05b7680-8303-11e9-9fdb-41c8402ba333.png)


![edit_globals_light_ 02](https://user-images.githubusercontent.com/12654295/58674218-a4879400-8303-11e9-83fe-03f484717273.png)


![edit_globals_light_ 03](https://user-images.githubusercontent.com/12654295/58674221-a6e9ee00-8303-11e9-813c-0fbfdf3c3e36.png)

bradymiller commented:
Up For Grabs demo for this PR is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Zeta_-_Up_For_Grabs_Demo

bradymiller commented:
Another big wow on this. And now we can add many more tabs :)

bradymiller commented on interface/super/edit_globals.php:
```diff
@@ -335,467 +335,563 @@ function checkBackgroundServices()
+                        <div class="input-group col-sm-4 oe-pull-away">
+                        <?php // mdsupport - Optional server based searching mechanism for large number of fields on this screen.
+                        if (!$userMode) {
+                            $placeholder = xlt('Search global settings');
```
change xlt to xl

bradymiller replied:
actually:
change xlt to xla

zbig01 replied:
will do

bradymiller commented on interface/super/edit_globals.php:
```diff
@@ -335,467 +335,563 @@ function checkBackgroundServices()
-                        echo "   <option value='" . attr($code_key) . "'";
-                        if ($code_key == $fldvalue) {
-                            echo " selected";
+                            $placeholder = xlt('Search user settings');
```
change xlt to xla

zbig01 replied:
ditto

bradymiller commented on interface/super/edit_globals.php:
```diff
@@ -335,467 +335,563 @@ function checkBackgroundServices()
+
+                                    echo "<div class=''>";
+                                    $addendum = $grpname == 'Appearance' ? ' (*'. xl("need to logout/login after changing these settings") .')' : '';
+                                    echo "<div class='col-sm-12 oe-global-tab-heading'><div class='oe-pull-toward' style='font-size: 1.4em'>". xl($grpname) ." &nbsp;</div><div style='margin-top: 5px'>" . $addendum ."</div></div>";
```
use xlt() around $grpname
use text() around $addendum

zbig01 replied:
done

bradymiller commented on interface/super/edit_globals.php:
```diff
@@ -335,467 +335,563 @@ function checkBackgroundServices()
+                                        }
+                                    }
+                                   
+                                    echo "<div><div class='oe-pull-away oe-margin-t-10' style=''>". xl($grpname) ." &nbsp;<i class='fa fa-lg fa-arrow-circle-up oe-help-redirect scroll' aria-hidden='true'></i></div><div class='clearfix'></div></div>";
```
use xlt() around $grpname

zbig01 replied:
done

bradymiller commented on interface/super/edit_globals.php:
```diff
@@ -335,467 +335,563 @@ function checkBackgroundServices()
+<?php
+$post_srch_desc = $_POST['srch_desc'];
+if (!empty($post_srch_desc) && $srch_item == 0) {
+    echo "<script>alert('{$post_srch_desc} - " . xlt('search term was not found, please try another search') . "');</script>";
```
change this to:
```
echo "<script>alert(" . js_escape($post_srch_desc." - ".xl('search term was not found, please try another search')) . ");</script>";
```

zbig01 replied:
Done

zbig01 commented:
@bradymiller 
Any more suggestions, if not I will update the PR

bradymiller commented:
No more suggestions and is testing well on my testing. Lets just wait for @adunsulag to ensure his issues were secondary to not building stylesheets, and then I'll bring this into the codebase.

sjpadgett commented:
Too lazy to setup and test but is the left menu fixed? If not, can it be?

zbig01 commented:
It is fixed :)
In a pinch can always check the demo :))

sjpadgett commented:
You just don't know how lazy I truly am. Thanks

zbig01 commented:
I am sorry that the screenshots were not explicit enough :)

sjpadgett commented:
Well remember your audience in the future my friend. Anyway, nice job and i'm off.

adunsulag commented:
@zbig01 The npm run build had to be done inside the docker container for the changes to work.  I've tested it and it is now working.  Tested on Chrome emulator for ipad,ipad+pro, iPhone 5,6,7,8, galaxy s5, and every desktop responsive width from 50px - 3200px.  Looks good.  Minor thing but the miscellaneous tab has a few letters that gets cut off at about 350px width.

adunsulag commented on interface/themes/colors/openemr5/edit_globals_colors.scss:
```diff
@@ -0,0 +1,35 @@
+/*vertical menu styling for edit globals - colors*/
+/*openrmr5.0.2 edit_globals vertical menu*/
+#oe-nav-ul.tabNav li.current {
+	color: #333;
```
Do we want this defined in a variable instead of a hard-coded value of #333?

adunsulag commented on interface/themes/colors/openemr5/edit_globals_colors.scss:
```diff
@@ -0,0 +1,35 @@
+    border-left: none;
+	border-bottom: none;
+	background: $pale;
+	color: #333;
```
Here is that color of 333 again.  Probably needs to be a defined SCSS variable.

adunsulag commented on interface/themes/colors/openemr5/edit_globals_colors.scss:
```diff
@@ -0,0 +1,35 @@
+	border-color: $dark;
+}
+.oe-global-tab-heading {
+	color: #000;
```
Can't we use the $black variable here?

adunsulag commented on interface/themes/colors/style_ash_blue.scss:
```diff
@@ -9,7 +9,7 @@ $pale: #B5BDDA;
 $midpalepale: #C3CAE1;
 $palepale: #E1E4F0;
 $palepurple: #E7E6F9;
-$paler: #F0F1F7;
```
This is minor, but in the commit description I would include the fact that you updated a number of the colors for the themes just so people know better what this commit did.  The current description doesn't state that the themes were changed as part of this.

adunsulag commented on interface/themes/core/oe-mobile.scss:
```diff
@@ -0,0 +1,34 @@
+/*for Smartphones*/
+@media only screen and (max-width: 400px) {
+    h1{
+        font-size: 30px;
```
Does the system have a global font-size definition at body?  Can we use rem here for each of these h* tags?  Just a thought as it makes it easier to override from the body and everything naturally adapts.  Its a more fundamental question for all the stylesheets but it helps avoid hard-coding pixel sizes here.

adunsulag commented on interface/themes/style_manila.scss:
```diff
@@ -464,13 +464,57 @@ body dl {
 /*~~openemr5 tabs menu , user and logout icon~~*/
 #menu_icon, #logout_icon, #user_icon {
-    color:#676666 !important;
+    color:#9c9c9c !important;
```
You use #9c9c9c in a lot of places in this file.  Should probably be pulled into a variable.

zbig01 commented:
@adunsulag 

Sorry for not having a more verbose commit message. In the past between the screenshots and the Demo the issue would not have arisen, however in this case you managed to check out the PR before the demo could be set up :)

I find it useful to run the following commands every time on my test OpenEMR instance before I install it. 

```
composer install
composer dump-autoload -o
npm install
npm audit fix
npm run build

```
It has saved me from grief many times.

Thank you for your thoughtful comments, you raise 2 important points - using the power of SASS and assigining variables to the colors and to peg the H* sizes to em's. In this instance I have narrowly tailored the styling to be applicable only to `edit_globals.php` so as to not mess with styling affecting other pages where the tab classes are being used. Going down that path would have led to making adjustments to several other scripts and I did not have the time to do that. Having said that I agree with you about using variables whenever possible. 

With regards the hard coded values in `style_manila.scss`, I was thinking of requesting the powers that be to drop style manila altogether as it is increasingly difficult to support that stylesheet. I therefore left the values as hard coded values in order to bring some consistency to the edit_globals page look when this stylesheet was used. In order to bring style manila up to snuff a lot of tweaking has to be done, that in and of itself is an immense task.

I too had the thought of linking the H* font-sizes to em's - a la Bootstrap 4 which has a base font of 16px. The math dictates an increase in the font-sizes for the various H* elements which I felt would make our headings look too big, I therfore decided to hard code the H* font-sizes for the mobile devices using a media query for displays less than 400 px. This will not affect normal and tablet sized displays as long as we are on Bootstrap 3 and when we move on to Bootstrap 4 I would prefer that we use the old Bootstrap 3 H* sizes.

I hope I have addressed some of your concerns. Given the wide difference in how various scripts have been written, bridging the gap between old and new scripts and maintaining a consisteny in the UI look and behavior is a delicate balancing act. I view this process as a gradual convergence in styling till we modernize all outward facing scripts.

adunsulag commented:
@zbig01 Thanks for your detailed response.  I think maintaining consistency with the rest of the themeing is an important point with a gradual move to more sanity.  I lack some of the historical knowledge right now of the overall context / consistency in the system so you have to take my suggestions with a grain of salt.  At least from my point in the review it looks good.

bradymiller commented:
hi @zbig01 , Thanks for yet another awesome code improvement; I honestly can't wait to see what you do next. And thanks for the code reviews @adunsulag and @sjpadgett ! From the comments above, it appears we are ready to bring in this PR, so here it goes :)
